File: frontend/src/lib/api/pricing.ts
/**
 * Pricing API Module
 * Pricing rules and calculation endpoints
 */

import { fetchWithAuth } from './client';

// ============================================================================
// Types
// ============================================================================

export interface PricingRule {
  id: string;
  name: string;
  description?: string;
  category: string;
  brand?: string;
  model_pattern?: string;
  age_min: number;
  age_max?: number;
  base_price: number;
  grade_modifiers: Record<string, number>;
  priority: number;
  is_active: boolean;
  created_at?: string;
  updated_at?: string;
}

export interface ConditionModifier {
  id: string;
  condition_name: string;
  display_name: string;
  description?: string;
  modifier: number;
  sort_order: number;
  is_active: boolean;
}

export interface PricingConfig {
  pricing_rules: PricingRule[];
  condition_modifiers: ConditionModifier[];
  categories: string[];
  brands: string[];
}

export interface PriceCalculationRequest {
  category: string;
  brand?: string;
  model?: string;
  age_years: number;
  grade: string;
  condition?: string;
}

export interface PriceCalculationResponse {
  base_price: number;
  grade_modifier: number;
  condition_modifier: number;
  final_price: number;
  pricing_rule_id?: string;
  pricing_rule_name?: string;
  breakdown: Record<string, unknown>;
}

// ============================================================================
// API
// ============================================================================

export const pricingApi = {
  getConfig: () =>
    fetchWithAuth<PricingConfig>('/pricing'),

  listRules: (params?: { category?: string; brand?: string; is_active?: boolean; skip?: number; limit?: number }) => {
    const query = new URLSearchParams();
    if (params) {
      if (params.category) query.append('category', params.category);
      if (params.brand) query.append('brand', params.brand);
      if (params.is_active !== undefined) query.append('is_active', String(params.is_active));
      if (params.skip !== undefined) query.append('skip', String(params.skip));
      if (params.limit !== undefined) query.append('limit', String(params.limit));
    }
    const qs = query.toString();
    return fetchWithAuth<PricingRule[]>(`/pricing/rules${qs ? `?${qs}` : ''}`);
  },

  getRule: (id: string) =>
    fetchWithAuth<PricingRule>(`/pricing/rules/${id}`),

  createRule: (data: Omit<PricingRule, 'id' | 'created_at' | 'updated_at'>) =>
    fetchWithAuth<PricingRule>('/pricing/rules', { method: 'POST', body: JSON.stringify(data) }),

  updateRule: (id: string, data: Partial<PricingRule>) =>
    fetchWithAuth<PricingRule>(`/pricing/rules/${id}`, { method: 'PUT', body: JSON.stringify(data) }),

  deleteRule: (id: string) =>
    fetchWithAuth<void>(`/pricing/rules/${id}`, { method: 'DELETE' }),

  listModifiers: (activeOnly = true) =>
    fetchWithAuth<ConditionModifier[]>(`/pricing/modifiers?active_only=${activeOnly}`),

  updateModifier: (id: string, data: Partial<ConditionModifier>) =>
    fetchWithAuth<ConditionModifier>(`/pricing/modifiers/${id}`, { method: 'PUT', body: JSON.stringify(data) }),

  calculatePrice: (request: PriceCalculationRequest) =>
    fetchWithAuth<PriceCalculationResponse>('/pricing/calculate', { method: 'POST', body: JSON.stringify(request) }),
};

File: backend/app/schemas/pricing.py
"""Pydantic schemas for Pricing API"""

from datetime import datetime
from typing import Optional, Dict, Any, List
from pydantic import BaseModel, Field, ConfigDict


# ============================================================================
# Pricing Rule Schemas
# ============================================================================

class PricingRuleBase(BaseModel):
    """Base schema for pricing rules"""
    name: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None
    category: str = Field(..., min_length=1, max_length=50)
    brand: Optional[str] = Field(None, max_length=100)
    model_pattern: Optional[str] = Field(None, max_length=255)
    age_min: int = Field(0, ge=0)
    age_max: Optional[int] = Field(None, ge=0)
    base_price: float = Field(..., gt=0)
    grade_modifiers: Dict[str, float] = Field(
        default_factory=lambda: {"A+": 1.0, "A": 0.85, "B": 0.70, "C": 0.50, "D": 0.30}
    )
    priority: int = Field(0, ge=0)
    is_active: bool = True


class PricingRuleCreate(PricingRuleBase):
    """Schema for creating a pricing rule"""
    pass


class PricingRuleUpdate(BaseModel):
    """Schema for updating a pricing rule"""
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    category: Optional[str] = Field(None, min_length=1, max_length=50)
    brand: Optional[str] = Field(None, max_length=100)
    model_pattern: Optional[str] = Field(None, max_length=255)
    age_min: Optional[int] = Field(None, ge=0)
    age_max: Optional[int] = Field(None, ge=0)
    base_price: Optional[float] = Field(None, gt=0)
    grade_modifiers: Optional[Dict[str, float]] = None
    priority: Optional[int] = Field(None, ge=0)
    is_active: Optional[bool] = None


class PricingRuleResponse(PricingRuleBase):
    """Schema for pricing rule response"""
    id: str
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    created_by: Optional[str] = None
    
    model_config = ConfigDict(from_attributes=True)


# ============================================================================
# Condition Modifier Schemas
# ============================================================================

class ConditionModifierBase(BaseModel):
    """Base schema for condition modifiers"""
    condition_name: str = Field(..., min_length=1, max_length=50)
    display_name: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = None
    modifier: float = Field(..., gt=0, le=2.0)
    sort_order: int = Field(0, ge=0)
    is_active: bool = True


class ConditionModifierCreate(ConditionModifierBase):
    """Schema for creating a condition modifier"""
    pass


class ConditionModifierUpdate(BaseModel):
    """Schema for updating a condition modifier"""
    condition_name: Optional[str] = Field(None, min_length=1, max_length=50)
    display_name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = None
    modifier: Optional[float] = Field(None, gt=0, le=2.0)
    sort_order: Optional[int] = Field(None, ge=0)
    is_active: Optional[bool] = None


class ConditionModifierResponse(ConditionModifierBase):
    """Schema for condition modifier response"""
    id: str
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    
    model_config = ConfigDict(from_attributes=True)


# ============================================================================
# Price Calculation Schemas
# ============================================================================

class PriceCalculationRequest(BaseModel):
    """Request to calculate price for a device"""
    category: str
    brand: Optional[str] = None
    model: Optional[str] = None
    age_years: float = Field(..., ge=0)
    grade: str = Field(..., description="Device grade: A+, A, B, C, D")
    condition: Optional[str] = Field(None, description="Condition: excellent, good, fair, poor")


class PriceCalculationResponse(BaseModel):
    """Response with calculated price"""
    base_price: float
    grade_modifier: float
    condition_modifier: float
    final_price: float
    pricing_rule_id: Optional[str] = None
    pricing_rule_name: Optional[str] = None
    breakdown: Dict[str, Any] = Field(default_factory=dict)


# ============================================================================
# Pricing Configuration Response
# ============================================================================

class PricingConfigResponse(BaseModel):
    """Complete pricing configuration"""
    pricing_rules: List[PricingRuleResponse]
    condition_modifiers: List[ConditionModifierResponse]
    categories: List[str]
    brands: List[str]

